// @feathersjs/polka/lib/index.js
const polka = require('polka');
const Proto = require('uberproto');
const errorHandler = require('@feathersjs/errors/handler');
const notFound = require('@feathersjs/errors/not-found');
const debug = require('debug')('@feathersjs/polka');

const rest = require('./rest');

function feathersPolka (feathersApp) {
  if (!feathersApp) {
    return polka();
  }

  if (typeof feathersApp.setup !== 'function') {
    throw new Error('@feathersjs/polka requires a valid Feathers application instance');
  }

  if (!feathersApp.version || feathersApp.version < '3.0.0') {
    throw new Error(`@feathersjs/polka requires an instance of a Feathers application version 3.x or later (got ${feathersApp.version || 'unknown'})`);
  }

  const polkaApp = polka();
  // An Uberproto mixin that provides the extended functionality
  const mixin = {
    register (location) {
      let service;
      let middleware = Array.from(arguments)
        .slice(1)
        .reduce(function (middleware, arg) {
          if (typeof arg === 'function') {
            middleware[service ? 'after' : 'before'].push(arg);
          } else if (!service) {
            service = arg;
          } else {
            throw new Error('Invalid options passed to app.use');
          }
          return middleware;
        }, {
          before: [],
          after: []
        });

      const hasMethod = methods => methods.some(name =>
        (service && !Array.isArray(service) && typeof service[name] === 'function')
      );

      // Check for service (any object with at least one service method)
      if (hasMethod(['handle', 'set']) || !hasMethod(this.methods.concat('setup'))) {
        debug('Passing app.register call to Polka app');
        return this._super.apply(this, arguments);
      }

      debug('Registering service with middleware', middleware);
      // Since this is a serivce, call Feathers `.register`
      feathersApp.use.call(this, location, service, { middleware });

      return this;
    },

    listen () {
      const server = this._super.apply(this, arguments);

      this.setup(server);
      debug('Feathers application listening');

      return server;
    }
  };

  // Copy all non-existing properties (including non-enumerables)
  // that don't already exist on the Polka app
  Object.getOwnPropertyNames(feathersApp).forEach(prop => {
    const feathersProp = Object.getOwnPropertyDescriptor(feathersApp, prop);
    const polkaProp = Object.getOwnPropertyDescriptor(polkaApp, prop);

    if (polkaProp === undefined && feathersProp !== undefined) {
      Object.defineProperty(polkaApp, prop, feathersProp);
    }
  });

  return Proto.mixin(mixin, polkaApp);
}

module.exports = feathersPolka;

Object.assign(module.exports, polka, {
  default: feathersPolka,
  original: polka,
  rest,
  notFound,
  errorHandler
});
